ckks/encoder.go

```diff
@@ -2,858 +2,1218 @@ package ckks
+func (ecd *Encoder) polyToFloatCRT(p ring.Poly, values FloatSlice, scale rlwe.Scale, logSlots int, r *ring.Ring) (err error) {
+
+	var slots int
+	switch values := values.(type) {
+	case []float64:
+		slots = utils.Min(len(p.Coeffs[0]), len(values))
+	case []complex128:
+		slots = utils.Min(len(p.Coeffs[0]), len(values))
+	case []*big.Float:
+		slots = utils.Min(len(p.Coeffs[0]), len(values))
+	case []*bignum.Complex:
+		slots = utils.Min(len(p.Coeffs[0]), len(values))
+	}
+
+	bigintCoeffs := ecd.bigintCoeffs
+
+	ecd.parameters.RingQ().PolyToBigint(ecd.buff, 1, bigintCoeffs)
 
-	values := make([]*ring.Complex, ecd.m>>2)
-	valuesfloat := make([]*big.Float, ecd.m>>1)
+	Q := r.ModulusAtLevel[r.Level()]
 
-	for i := 0; i < ecd.m>>2; i++ {
+	ecd.qHalf.Set(Q)
+	ecd.qHalf.Rsh(ecd.qHalf, 1)
+
+	var sign int
+	for i := 0; i < slots; i++ {
+		// Centers the value around the current modulus
+		bigintCoeffs[i].Mod(bigintCoeffs[i], Q)
 
-		values[i] = ring.NewComplex(ring.NewFloat(0, ecd.prec), ring.NewFloat(0, ecd.prec))
-		valuesfloat[i*2] = ring.NewFloat(0, ecd.prec)
-		valuesfloat[(i*2)+1] = ring.NewFloat(0, ecd.prec)
+		sign = bigintCoeffs[i].Cmp(ecd.qHalf)
+		if sign == 1 || sign == 0 {
+			bigintCoeffs[i].Sub(bigintCoeffs[i], Q)
+		}
 	}
 
-	return &encoderBigComplex{
-		encoder:     *ecd.encoder.ShallowCopy(),
-		zero:        ring.NewFloat(0, ecd.prec),
-		cMul:        ring.NewComplexMultiplier(),
-		prec:        ecd.prec,
-		values:      values,
-		valuesfloat: valuesfloat,
-		roots:       ecd.roots,
+	switch values := values.(type) {
+
+	case []float64:
+		sf64 := scale.Float64()
+		for i := 0; i < slots; i++ {
+			values[i] = scaleDown(bigintCoeffs[i], sf64)
+		}
+	case []complex128:
+		sf64 := scale.Float64()
+		for i := 0; i < slots; i++ {
+			values[i] = complex(scaleDown(bigintCoeffs[i], sf64), 0)
+		}
+	case []*big.Float:
+		s := &scale.Value
+		for i := 0; i < slots; i++ {
+
+			if values[i] == nil {
+				values[i] = new(big.Float)
+			}
+
+			values[i].SetInt(bigintCoeffs[i])
+			values[i].Quo(values[i], s)
+		}
+	case []*bignum.Complex:
+		s := &scale.Value
+		for i := 0; i < slots; i++ {
+
+			if values[i] == nil {
+				values[i] = &bignum.Complex{
+					new(big.Float),
+					new(big.Float),
+				}
+			} else {
+				if values[i][0] == nil {
+					values[i][0] = new(big.Float)
+				}
+			}
+
+			values[i][0].SetInt(bigintCoeffs[i])
+			values[i][0].Quo(values[i][0], s)
+		}
+	default:
+		return fmt.Errorf("cannot polyToComplexNoCRT: values.(Type) must be []complex128, []*bignum.Complex, []float64 or []*big.Float but is %T", values)
 	}
+
+	return
 }
 
-func (ecd *encoderBigComplex) decodePublic(plaintext *rlwe.Plaintext, logSlots int, sigma float64) (res []*ring.Complex) {
+// polyToFloatNoCRT decodes a single-level CRT poly on a real valued FloatSlice.
+func (ecd *Encoder) polyToFloatNoCRT(coeffs []uint64, values FloatSlice, scale rlwe.Scale, logSlots int, r *ring.Ring) (err error) {
```

Comment:
This function isn't tested in `ckks_test.go` it might be good to have a case what uses it.

Reply:
Done